File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N];
        int idx = 0;
        int num = 1;
        int cnt = 0;
        Stack<Integer> stack = new Stack<>();
        for(int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        while(idx < N && cnt <= N * 2) {
            if(!stack.isEmpty() && stack.peek() == arr[idx]) {
                stack.pop();
                sb.append('-').append('\n');
                idx++;
            } else {
                stack.push(num++);
                sb.append('+').append('\n');
            }
            cnt++;
        }
        System.out.println(idx == N ? sb.toString().trim() : "NO");
    }
}

File: 백준/Silver/1874. 스택 수열/스택 수열.js
class Stack {
  constructor() {
    this.arr = [];
  }

  push(num) {
    this.arr.push(num);
  }

  pop() {
    return this.arr.length === 0 ? -1 : this.arr.pop();
  }

  size() {
    return this.arr.length;
  }

  isEmpty() {
    return this.arr.length === 0 ? true : false;
  }

  peek() {
    return this.arr.length === 0 ? -1 : this.arr[this.arr.length - 1];
  }
}

const [N, ...input] = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .split("\n")
  .map((line) => Number(line.trim()));
let stack = new Stack();
let cnt = 0;
let idx = 0;
let num = 1;
const ans = [];
while (idx < N && cnt <= N * 2) {
  if (!stack.isEmpty() && stack.peek() === input[idx]) {
    stack.pop();
    ans.push("-");
    idx++;
  } else {
    stack.push(num++);
    ans.push("+");
  }
  cnt++;
}
console.log(idx === N ? ans.join("\n") : "NO");
